RIGS/views/finance.py

```diff
@@ -18,8 +18,76 @@
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context['page_title'] = "Invoice Dashboard"
+        context['description'] = "Overview of financial status of TEC rigs."
+
+        time_filter = self.request.GET.get('time_filter', 'all')
+
+        if time_filter not in TIME_FILTERS:
+            time_filter = 'all'
+
+        if time_filter == 'all':
+            context['events'] = models.Event.objects.all()
+            context['invoices'] = models.Invoice.objects.all()
+            context['payments'] = models.Payment.objects.all()
+        elif time_filter == 'year':
+            context['events'] = models.Event.objects.filter(start_date__gte=datetime.date.today() - datetime.timedelta(days=365))
+            context['invoices'] = models.Invoice.objects.filter(invoice_date__gte=datetime.date.today() - datetime.timedelta(days=365))
+            context['payments'] = models.Payment.objects.filter(date__gte=datetime.date.today() - datetime.timedelta(days=365))
+        elif time_filter == 'month':
+            context['events'] = models.Event.objects.filter(start_date__gte=datetime.date.today() - datetime.timedelta(days=30))
+            context['invoices'] = models.Invoice.objects.filter(invoice_date__gte=datetime.date.today() - datetime.timedelta(days=30))
+            context['payments'] = models.Payment.objects.filter(date__gte=datetime.date.today() - datetime.timedelta(days=30))
+        elif time_filter == 'week':
+            context['events'] = models.Event.objects.filter(start_date__gte=datetime.date.today() - datetime.timedelta(days=7))
+            context['invoices'] = models.Invoice.objects.filter(invoice_date__gte=datetime.date.today() - datetime.timedelta(days=7))
+            context['payments'] = models.Payment.objects.filter(date__gte=datetime.date.today() - datetime.timedelta(days=7))
+
+        context["time_filter"] = time_filter
+
+        context['total_outstanding'] = sum([i.balance for i in models.Invoice.objects.outstanding_invoices()])
+        context['total_waiting'] = sum([i.sum_total for i in models.Event.objects.waiting_invoices()])
+        context['total_events'] = len(context['events'])
+        context['total_invoices'] = len(context['invoices'])
+        context['total_payments'] = len(context['payments'])
+
+        payment_methods = dict(models.Payment.METHODS)
+
+        context['payment_methods'] = context["payments"].values('method').annotate(total=Sum('amount')).order_by('method')
+
+        for method in context['payment_methods']:
+            method['method'] = payment_methods.get(method['method'], f"Unknown method ({method['method']})")
+
+        context["total_income"] = sum([i['total'] for i in context['payment_methods']])
+
+        payments = context['payments']
+        mean_duration = 0
+
+        for payment in payments:
+            mean_duration += days_between(payment.invoice.invoice_date, payment.date)
```

Comment:
Can you not just subtract invoice_date from payment_date and then do .days on the timedelta? Why does it need to be converted to seconds?

Reply:
This method allows for fractional days to be counted, I agree we could do that and get the same result but this method is more accurate and for testing purposes works a bit nicer.

The `days` property is always an integer, not a float.

Reply:
Ah I'm with you now :+1: